### PR TITLE
102 refactor to typescript rebased

### PR DESCRIPTION
**Pull request by hoelger**

### CODE REVIEW

**Comment by GeorgHoelger** [comment]
I ran npm run-script lint-fix and npm run-script postlint-fix, the first one also changed some styling. But when I now change some indentation in /app then those commands don't revert these changes. What? My best proposal for this is:

    Extract this in some follow-up issue where:
    The procedure/commands are documented in the README and
    Once you set up everything correctly (e.g. with pre-commit) you're acutally forced to comply to the styling rules

**Comment by GeorgHoelger** [comment]
Just checked: All files containing `[extends] Component<` have the `.tsx` extension

File: app/actions/bot.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
+import { $Values } from "utility-types";
```
Comment (neopostmodern) [inline app/actions/bot.ts:1]: This seems strange. Is there really no built-in TypeScript way of solving whatever this does?

File: app/actions/cache.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+export function resetCache(): Action {
+  return {
+    type: RESET_CACHE,
+    payload: null,
+    meta: {
+      cache: true
+    }
+  };
+}
+export function markCompleted(name: string, identifier: string, // eslint-disable-next-line flowtype/no-weak-types
```
Comment (neopostmodern) [inline app/actions/cache.ts:21]: all `eslint-disable-next-line flowtype/*` should be removed, if necessary to be replaced by equivalent typescript rules

File: app/Routes.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
```
Comment (neopostmodern) [inline app/Routes.tsx:1]: Do you understand why some files are marked as deleted/created and others are recognized as renamed?

File: app/actions/electron.ts [hunk 4]
```diff
@@ -1,24 +1,7 @@
...
-  SHOW_DEV_TOOLS,
-  OPEN_PDF,
-  SET_AVAILABLE_VERSION,
-  SET_UPDATE_DOWNLOAD_PROGRESS
-} from '../constants/actionTypes';
-
+import { MAIN } from "../constants/targets";
+import type { Action } from "../reducers/types";
+import type { BrowserViewName } from "../reducers/electron";
+import { ELECTRON_ROUTING, HIDE_CONFIGURATION, INTERNAL_ADD_BROWSER_VIEW, PERFORM_SCROLL, SET_BROWSER_VIEW_READY, SET_BROWSER_VIEW_URL, SET_BROWSER_WINDOW, SET_INTERACTIVE_MODE, SHOW_CONFIGURATION, SHOW_DEV_TOOLS, OPEN_PDF, SET_AVAILABLE_VERSION, SET_UPDATE_DOWNLOAD_PROGRESS } from "../constants/actionTypes";
```
Comment (neopostmodern) [inline app/actions/electron.ts:4]: is this super-long line the output of prettier?

File: app/actions/email.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
+import { SEND_MAIL } from "../constants/actionTypes";
+import { MAIN } from "../constants/targets";
+import { generateInPlaceDescription } from "../flat/applicationTextBuilder";
+import { flatPageUrl } from "../flat/urlBuilder";
+import type { ContactData } from "../reducers/configuration";
+import type { OverviewDataEntry } from "../reducers/data";
+import type { Action, Dispatch, GetState } from "../reducers/types";
+import { markCompleted } from "./cache";
+import { CACHE_NAMES } from "../reducers/cache";
+// eslint-disable-next-line import/prefer-default-export
```
Comment (neopostmodern) [inline app/actions/email.ts:10]: pretty sure this `esline-disable-next-line` is unnecessary, could you check if we have "no unnecessary rule disabling" active and if yes, why this didn't yield an error?

File: app/actions/cache.ts (hunk 2 above)
Comment (neopostmodern) [reply app/actions/cache.ts:21]: somewhat related: do we have any (new) typescript linting rules?

File: app/components/BotOverlay.tsx [hunk 6]
```diff
@@ -0,0 +1,132 @@
+import React, { Component } from "react";
+import styles from "./BotOverlay.scss";
+// @ts-expect-error - flow doesn't like SVG
```
Comment (neopostmodern) [inline app/components/BotOverlay.tsx:3]: should be obsolete?

File: app/components/BotOverlay.tsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    if (isPuppetLoading) {
+      return `Website lädt...`;
+    }
+
+    if (botMessage) {
+      return botMessage;
+    }
+
+    // eslint-disable-next-line flowtype/no-weak-types
+    const matchedFlats = ((Object.values(verdicts) as any) as Array<Verdict>).filter(({
```
Comment (neopostmodern) [inline app/components/BotOverlay.tsx:79]: strange double-cast, should be possible without any cast at all!

File: app/components/BotOverlay.tsx [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    const speechBubbleText = this.botTalk();
+    return <div className={styles.container} data-tid="container" onWheel={this.handleWheel}>
+        {
+        /* make sure to draw the bottom and top border atop of the overlays */
+      }
+        <div className={styles.borderBottom} />
+        <div className={styles.borderTop} />
+
+        {BotOverlay.renderAnimations(animations)}
+        {showOverlay ? overviewBoundingBoxes.map(({
```
Comment (neopostmodern) [inline app/components/BotOverlay.tsx:113]: this also feels like prettier did not do it's job? very weird ternary formatting, but i might be wrong.

File: app/components/Configuration.tsx [hunk 9]
```diff
@@ -1,49 +1,43 @@
...
-  resetConfiguration: () => void,
-  configuration: ConfigurationType
+  nextStage: () => void;
+  previousStage: () => void;
+  hideConfiguration: () => void;
+  togglePostcode: (postcode: string) => void;
+  resetPostcodes: () => void;
+  toggleFloor: (floor: number) => void;
+  toggleBoolean: (name: configurationBoolean) => void;
+  setNumber: (name: configurationNumbers, value: number | null | undefined) => void;
```
Comment (neopostmodern) [inline app/components/Configuration.tsx:17]: i'm confused that both `null` and `undefined` are allowed here, have you checked if this makes sense where this call ends up?

File: app/components/Configuration.tsx [hunk 10]
```diff
@@ -164,31 +160,17 @@ export default class Configuration extends Component<Props> {
           </div>
 
           <div className={styles.header}>
-            {configuration.stage > 1 &&
-            configuration.stage < stages.length - 1 ? (
-              <>
+            {configuration.stage > 1 && configuration.stage < stages.length - 1 ? <>
```
Comment (neopostmodern) [inline app/components/Configuration.tsx:163]: i like the old formatting much better, really wondering if prettier is set up correctly

File: app/components/PostcodeMap.tsx [hunk 11]
```diff
@@ -1,76 +1,69 @@
...
-import { divIcon, type Layer } from 'leaflet';
-import { feature } from 'topojson';
-import topoData from '../map/berlin-postcodes-data.topo';
-import labels from '../map/labels';
-
+import type { Layer } from "leaflet";
+import { divIcon } from "leaflet";
+import { feature } from "topojson";
+import topoData from "../map/berlin-postcodes-data.topo";
+import labels from "../map/labels";
```
Comment (neopostmodern) [inline app/components/PostcodeMap.tsx:11]: newline after imports

File: app/components/configurationStages/personalData.tsx [hunk 12]
```diff
@@ -1,31 +1,33 @@
...
+import React from "react";
+import styles from "../Configuration.scss";
+import { EMPLOYMENT_STATUS, SALUTATIONS, USEACCOUNT } from "../../reducers/configuration";
+import type { Configuration } from "../../reducers/configuration";
+import type { InheritedProps, StageDescription } from "./types";
+import Disclaimer from "./disclaimer";
+import TextField from "../inputs/TextInput";
+import NumberField from "../inputs/NumberField";
+import EnumField from "../inputs/EnumField";
+import YesNo from "../inputs/YesNo";
```
Comment (neopostmodern) [inline app/components/configurationStages/personalData.tsx:10]: confused that the formatter *removed* the newline after imports

File: app/components/configurationStages/review.tsx [hunk 13]
```diff
@@ -61,190 +58,143 @@ const reviewStage: StageDescription = {
...
-          {/*    <em>keine</em>{' '} */}
-          {/*  </> */}
-          {/* )} */}
-          {/* anonymen Daten zur Wohnungsnot sammeln */}
-          <br />
-          <br />
-          {configuration.experimentalFeatures.sortByNewest ? (
-            <>
+            </>}
+          {
```
Comment (neopostmodern) [inline app/components/configurationStages/review.tsx:156]: weird formatting, but could also simply be removed

File: app/middleware/data.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        overview,
+        flat,
+        verdicts
+      },
+      configuration
+    } = store.getState();
+
+    if (overview) {
+      const configurationHash = getConfigurationFilterHash(configuration);
+      // $FlowFixMe -- Object.values
```
Comment (neopostmodern) [inline app/middleware/data.ts:22]: can't remember what `$FlowFixMe` did, either convert to it's TypeScript equivalent or remove

File: app/middleware/helpers.ts [hunk 15]
```diff
@@ -1,25 +1,26 @@
...
+import sendMail from "../utils/email";
+import { sleep, timeout } from "../utils/async";
+import { electronRouting } from "../actions/electron";
+import { electronObjects } from "../store/electronObjects";
 const pdfFolderPath = path.join(app.getPath('userData'), 'pdf');
+
 if (!fs.existsSync(pdfFolderPath)) {
   fs.mkdirSync(pdfFolderPath);
-}
+} // eslint-disable-next-line no-unused-vars
```
Comment (neopostmodern) [inline app/middleware/helpers.ts:14]: eslint rule moved to wrong line

Comment (neopostmodern) [reply app/middleware/helpers.ts:14]: this type of error should be caught by eslint if we (re)enabled "no unnecessary disable ..."

File: app/store/electronObjects.ts [hunk 16]
```diff
@@ -1,4 +1,4 @@
 export const electronObjects = {
   views: {},
   window: null
-};
+};
```
Comment (neopostmodern) [inline app/store/electronObjects.ts:4]: i also find it strange that terminating newlines are removed, i thought they were best practice still

File: app/utils/randomUserAgent.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
+const operatingSystemIdentifiers = ['Windows NT 10.0; Win64; x64', 'Windows NT 6.3; Win64; x64', 'Windows NT 6.2; Win64; x64', 'Windows NT 6.1; Win64; x64', 'Windows NT 10.0; WOW64', 'Macintosh; Intel Mac OS X 10_14_5', 'Macintosh; Intel Mac OS X 10_14_4', 'Macintosh; Intel Mac OS X 10_13_6'];
+const chromeVersionIdentifiers = ['74.0.3729.169', '72.0.3626.121', '70.0.3538.102', '60.0.3112.90'];
```
Comment (neopostmodern) [inline app/utils/randomUserAgent.ts:2]: whoa, this might be the cause of #184 !

File: internals/scripts/CheckBuiltsExist.js
Comment (neopostmodern) [inline internals/scripts/CheckBuiltsExist.js:1]: empty file?

**Review by neopostmodern — CHANGES_REQUESTED** [review]
i only skimmed through most of the code and would say that as long as it's still working, everything should be fine.
the formatting things i commented throughout really irritate me, i feel like something is off there.
please remove all the flow specific comments.

**Comment by neopostmodern** [comment]
Also, the build fails for macOS - might be completely unrelated to the TypeScript thing, let's see if other builds get through: 
```
Error: Cannot find module 'dmg-license'
```

**Comment by GeorgHoelger** [comment]
> Also, the build fails for macOS - might be completely unrelated to the TypeScript thing, let's see if other builds get through:
> 
> ```
> Error: Cannot find module 'dmg-license'
> ```

From what I see only the macOS job fails. Linux always runs through, and Windows runs through when it's faster than the macOS. If it's slower than the macOS job, then somehow it get's canceled.

File: app/utils/randomUserAgent.ts (hunk 17 above)
Comment (GeorgHoelger) [reply app/utils/randomUserAgent.ts:2]: Mentioned in the issue, closing here

File: app/Routes.tsx (hunk 3 above)
Comment (GeorgHoelger) [reply app/Routes.tsx:1]: Nope. I think this is just the scratching the limits of git magic. 

Sometimes the git magic is depending on what program you use. For example `sublime merge` would display indentation changes very badly, but online in GitLab you could see everything really good.

I mean:
```
   if condition:
       <old code with one more indent>
   else:
        some new stuff here
```

File: app/actions/electron.ts (hunk 4 above)
Comment (GeorgHoelger) [reply app/actions/electron.ts:4]: Nope, its from the translation tool.

I created issue #187 which should take care of correct styling

File: internals/scripts/CheckBuiltsExist.js
Comment (GeorgHoelger) [reply internals/scripts/CheckBuiltsExist.js:1]: Ah yeah some rebase magic in combination with #151

File: app/actions/bot.ts (hunk 1 above)
Comment (GeorgHoelger) [reply app/actions/bot.ts:1]: I think TypeScript supports straight-forward notation if you use enums (from what I remember those don't really exist in JavaScript...) 

But the variables here (as `LOGINSTATUS`) are no enums, they're `const` that might be the reason for the cumbersome notation

Just a guess..

**Comment by GeorgHoelger** [comment]
Where would TypeScript errors actually would pop up? Is there also the option to ignore them (like the prettier styling) or can we enforce the TS rules? :)

File: app/actions/email.ts (hunk 5 above)
Comment (GeorgHoelger) [reply app/actions/email.ts:10]: removed the disable

could not find the "no unnecessary disabling" thingy...

File: app/components/PostcodeMap.tsx (hunk 11 above)
Comment (GeorgHoelger) [reply app/components/PostcodeMap.tsx:11]: noted down in the prettier issue

File: app/components/configurationStages/personalData.tsx (hunk 12 above)
Comment (GeorgHoelger) [reply app/components/configurationStages/personalData.tsx:10]: yes, that's the flow->ts formatter that did such things

File: app/middleware/data.ts (hunk 14 above)
Comment (GeorgHoelger) [reply app/middleware/data.ts:22]: First I removed all of them, then I found this. 

In general I feel weird: I have the feeling that doesn't matter what I do, the code just runs :D I have to get more used to the pipeline, understand where errors would pop up if they pop up.

File: app/actions/cache.ts (hunk 2 above)
Comment (GeorgHoelger) [reply app/actions/cache.ts]: I removed all flowtype disables and now running `npm run-script lint` there come up 28 errors but rather on config/build files so I think we don't have to replace them.

File: app/components/BotOverlay.tsx (hunk 7 above)
Comment (GeorgHoelger) [reply app/components/BotOverlay.tsx:75]: For me this looks like type-annotations were translated into type-casting... but maybe I'm wrong.

previous code was ` ((Object.values(verdicts): any): Array<Verdict>).filter(...)`

Removed the any-cast now

File: app/components/Configuration.tsx (hunk 9 above)
Comment (GeorgHoelger) [reply app/components/Configuration.tsx:17]: yes, for the setNumber there it should be only null as I see it from https://github.com/neopostmodern/wohnungsbot/blob/dev/app/components/inputs/NumberField.js#L3

for the setString (one line below) I'm confused and cannot judge if we can narrow it down

File: app/actions/bot.ts (hunk 1 above)
Comment (neopostmodern) [reply app/actions/bot.ts:1]: Could we transform them to enums though?

File: app/actions/email.ts (hunk 5 above)
Comment (neopostmodern) [reply app/actions/email.ts]: https://github.com/mysticatea/eslint-plugin-eslint-comments/blob/master/docs/rules/no-unused-disable.md
Not sure how easy it is to set up and if it's worth it, could you look into that?

File: app/actions/electron.ts (hunk 4 above)
Comment (neopostmodern) [reply app/actions/electron.ts:4]: #187 is good and important, but did you run prettier manually on the currently pushed code?

Comment (neopostmodern) [reply app/actions/electron.ts:4]: I.e. I think we should think of #187 and having a pretty PR here as separate concerns.

File: app/components/PostcodeMap.tsx (hunk 11 above)
Comment (neopostmodern) [reply app/components/PostcodeMap.tsx:11]: I'm not sure if prettier can *add* newlines in this case. The PR should not make the formatting worse, so we might need to tackle this manually.

File: app/store/electronObjects.ts (hunk 16 above)
Comment (neopostmodern) [reply app/store/electronObjects.ts:4]: Prettier will do this automatically: https://prettier.io/docs/en/rationale.html#empty-lines

**Comment by neopostmodern** [comment]
There is `// @ts-ignore` to disable type checking for the next line, but it should be checked whether explicit hacks (such as any-casting) might be more explicit.

**Comment by neopostmodern** [comment]
Concerning `dmg-license`... it's a mess: https://github.com/electron-userland/electron-builder/issues/6520#issuecomment-1436124454

**Review by neopostmodern — APPROVED** [review]

**Pull request merged**